File: functions/src/user/noteService.ts
/**
 * Copyright © 2020 WorldRIZe. All rights reserved.
 */
import { UserRepository } from './userRepository'
import { Note, NotePhrase } from './model/note'
import { v4 as uuidv4 } from 'uuid'
import _ from 'lodash'
import moment from 'moment'
import * as functions from 'firebase-functions'

export class NoteService {
  private readonly repo: UserRepository

  constructor(repo: UserRepository) {
    this.repo = repo
  }

  static generateNotePhrase(): NotePhrase {
    return {
      schemaVersion: 'v1',
      id: uuidv4(),
      japanese: '',
      english: '',
      createdAt: moment().toISOString(),
    }
  }

  static generateNote(noteId: string, title: string, isDefaultNote: boolean = false, isAchievedNote: boolean = false): Note {
    return {
      schemaVersion: 'v1',
      id: noteId,
      isDefaultNote: isDefaultNote,
      isAchievedNote: isAchievedNote,
      title: title,
      sortType: 'createdAt-',
      // 30 phrases
      phrases: [...Array(30)].map(_ => NoteService.generateNotePhrase()),
    }
  }

  /**
   * ノートを作成
   * @param userId
   * @param note
   */
  async createNote(userId: string, note: Note): Promise<Note> {
    const user = await this.repo.findById(userId)

    const userNoteLimit = 3 + (user.items['extra_note'] ?? 0)
    const noteCount = Object.keys(user.notes).length
    console.log(`limit: ${userNoteLimit}, notes: ${noteCount}`)
    if (userNoteLimit <= noteCount) {
      throw new functions.https.HttpsError('out-of-range', `note limit exceeded`)
    }

    const noteId = uuidv4()
    user.notes[noteId] = note
    await this.repo.update(user)
    return note
  }

  /**
   * ノートを更新
   * @param userId
   * @param note
   */
  async updateNote(userId: string, note: Note): Promise<Note> {
    const user = await this.repo.findById(userId)
    user.notes[note.id] = note
    await this.repo.update(user)
    return note
  }

  /**
   * ノートを削除
   * @param userId
   * @param noteId
   */
  async deleteNote(userId: string, noteId: string): Promise<void> {
    const user = await this.repo.findById(userId)
    delete user.notes[noteId]
    await this.repo.update(user)
  }
}

File: lib/usecase/note_service.dart
// Copyright © 2020 WorldRIZe. All rights reserved.

import 'package:data_classes/data_classes.dart';
import 'package:uuid/uuid.dart';
import 'package:wr_app/domain/note/model/note.dart';
import 'package:wr_app/domain/note/model/note_phrase.dart';
import 'package:wr_app/infrastructure/note/i_note_repository.dart';
import 'package:wr_app/domain/user/model/user.dart';
import 'package:wr_app/usecase/exceptions.dart';

class NoteService {
  NoteService({
    @required INoteRepository noteRepository,
  }) : _noteRepository = noteRepository;

  final INoteRepository _noteRepository;

  Note createDummyNote(String title, {bool isDefault = false}) {
    final phrases =
        List.generate(30, (index) => NotePhrase.dummy(id: Uuid().v4()));
    return Note(
      id: Uuid().v4(),
      title: title,
      isAchievedNote: false,
      isDefaultNote: isDefault,
      sortType: '',
      phrases: phrases,
    );
  }

  /// throw NoteLimitExceeded
  /// AchievedNote + 3
  Future<Note> createNote({
    @required User user,
    @required Note note,
  }) async {
    final notes = await getAllNotes(user: user);
    if (notes.length >= 4) {
      throw NoteLimitExceeded('user: ${user.uuid}');
    }
    return _noteRepository.createNote(user: user, note: note);
  }

  Future<void> deleteNote({
    @required User user,
    @required Note note,
  }) {
    return _noteRepository.deleteNote(user: user, note: note);
  }

  Future<void> updateNote({
    @required User user,
    @required Note note,
  }) {
    return _noteRepository.updateNote(user: user, note: note);
  }

  Future<Map<String, Note>> getAllNotes({@required User user}) {
    return _noteRepository.getAllNotes(user: user);
  }
}
